zwave_js_server/client.py

```diff
@@ -199,181 +181,119 @@ async def async_send_json_message(self, message: Dict[str, Any]) -> None:
+        assert self.client
 
-                if self._logger.isEnabledFor(logging.DEBUG):
-                    self._logger.debug("Received message:\n%s\n", pprint.pformat(msg))
+        loop = asyncio.get_running_loop()
 
-                msg_ = cast(dict, msg)
-                try:
-                    self.async_handle_message(msg_)
+        while not self.client.closed:
+            try:
+                msg = await self.client.receive()
+            except (
+                client_exceptions.WSServerHandshakeError,
+                client_exceptions.ClientError,
+            ) as err:
+                raise ConnectionFailed(err) from err
+
+            if msg.type in (WSMsgType.CLOSED, WSMsgType.CLOSING):
+                break
 
-                except InvalidState as err:
-                    disconnect_warn = f"Invalid state: {err}"
-                    await client.close()
-                    break
+            if msg.type == WSMsgType.ERROR:
+                raise ConnectionFailed()
 
-                except Exception:  # pylint: disable=broad-except
-                    self._logger.exception("Unexpected error handling %s", msg)
-                    break
+            if msg.type != WSMsgType.TEXT:
+                raise InvalidMessage(f"Received non-Text message: {msg.type}")
 
-        except client_exceptions.WSServerHandshakeError as err:
-            self._logger.warning("Unable to connect: %s", err)
+            try:
+                if len(msg.data) > SIZE_PARSE_JSON_EXECUTOR:
+                    msg = await loop.run_in_executor(None, msg.json)
+                else:
+                    msg = msg.json()
+            except ValueError as err:
+                raise InvalidMessage("Received invalid JSON.") from err
 
-        except client_exceptions.ClientError as err:
-            self._logger.warning("Unable to connect: %s", err)
+            if self._logger.isEnabledFor(logging.DEBUG):
+                self._logger.debug("Received message:\n%s\n", pprint.pformat(msg))
 
-        except asyncio.CancelledError:
-            pass
+            msg_ = cast(dict, msg)
 
-        finally:
-            if disconnect_warn is None:
-                self._logger.info("Connection closed")
-            else:
-                self._logger.warning("Connection closed: %s", disconnect_warn)
+            try:
+                self.async_handle_message(msg_)
+            except InvalidState:
+                await self._close()
+                raise
 
     async def disconnect(self) -> None:
         """Disconnect the client."""
-        self.close_requested = True
-
         if self.client is not None:
-            await self.client.close()
-        elif self.retry_task is not None:
-            self.retry_task.cancel()
+            await self._close()
+
+        self.state = STATE_DISCONNECTED
 
-        if self._disconnect_event is not None:
-            await self._disconnect_event.wait()
+        if self._on_disconnect:
```

Comment:
We now fire these only when we manually call disconnect. How will unexpected disconnect be notified?

Reply:
The `listen` method will raise an error in all cases except for `WSMsgType.CLOSED, WSMsgType.CLOSING` messages.

The docs are sparse on the message type topic. Is this ok?

Reply:
Just to be clear, the recommended way to use the library is now with the context manager interface. This will disconnect the client upon error.

I'll update the readme with this.

Reply:
We can block calling `listen` by setting a boolean in `aenter` and checking for that.

Reply:
That way we don't have to deal with it.

Reply:
But I also think that `listen` cancelling should disconnect the client. So the disconnect method just cancels that task.